plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange/CodeRange.hs

```diff
@@ -0,0 +1,109 @@
+{-# LANGUAGE DeriveGeneric             #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE NamedFieldPuns            #-}
+{-# LANGUAGE OverloadedStrings         #-}
+{-# LANGUAGE TypeFamilies              #-}
+
+module Ide.Plugin.SelectionRange.CodeRange
+    ( CodeRange (..)
+    , GetCodeRange(..)
+    , codeRangeRule
+    , Log
+    , useExcept
+    ) where
+
+import           Control.DeepSeq                         (NFData)
+import           Control.Monad.Except                    (ExceptT (..),
+                                                          runExceptT)
+import           Control.Monad.Reader                    (runReader)
+import           Control.Monad.Trans.Maybe               (MaybeT (MaybeT),
+                                                          maybeToExceptT)
+import           Data.Coerce                             (coerce)
+import           Data.Data                               (Typeable)
+import           Data.Hashable
+import qualified Data.Map.Strict                         as Map
+import           Development.IDE
+import           Development.IDE.Core.Rules              (toIdeResult)
+import qualified Development.IDE.Core.Shake              as Shake
+import           Development.IDE.GHC.Compat              (Annotated,
+                                                          HieAST (..),
+                                                          HieASTs (getAsts),
+                                                          ParsedSource, RefMap)
+import           Development.IDE.GHC.Compat.Util
+import           Development.IDE.GHC.ExactPrint          (GetAnnotatedParsedSource (GetAnnotatedParsedSource))
+import           GHC.Generics                            (Generic)
+import           Ide.Plugin.SelectionRange.ASTPreProcess (PreProcessEnv (..),
+                                                          preProcessAST)
+import           Prelude                                 hiding (log)
+
+data Log = LogShake Shake.Log
+    | LogBadDependency BadDependencyLog
+    | LogNoAST
+
+instance Pretty Log where
+    pretty log = case log of
+        LogShake shakeLog                 -> pretty shakeLog
+        LogBadDependency badDependencyLog -> pretty badDependencyLog
+        LogNoAST                          -> "no HieAst exist for file"
+
+data BadDependencyLog = forall rule. Show rule => BadDependencyLog rule
+
+instance Pretty BadDependencyLog where
+    pretty (BadDependencyLog rule) = "can not get result from rule " <> pretty (show rule)
+
+data CodeRange = CodeRange Range [CodeRange]
```

Comment:
haddock?

Reply:
Maybe record syntax so we can have documented fields?

Reply:
indeed